csharp/src/Ice/LoggerAdmin.cs

```diff
@@ -4,27 +4,99 @@
 
 using System;
 using System.Collections.Generic;
+using System.Threading.Tasks;
 using System.Diagnostics;
 using System.Linq;
 
 namespace ZeroC.Ice
 {
+    internal sealed class RemoteLoggerQueue
+    {
+        private readonly LoggerAdmin _loggerAdmin;
+        private readonly HashSet<LogMessageType> _messageTypes;
+        private readonly IRemoteLoggerPrx _remoteLoggerPrx;
+        private ValueTask _task;
+        private readonly HashSet<string> _traceCategories;
+
+        internal RemoteLoggerQueue(LoggerAdmin loggerAdmin, IRemoteLoggerPrx prx, HashSet<LogMessageType> messageTypes,
+            HashSet<string> traceCategories)
+        {
+            // Use oneway proxy to send the remote logger invocations.
+            _remoteLoggerPrx = prx.Clone(oneway: true);
+            _loggerAdmin = loggerAdmin;
+            _messageTypes = messageTypes;
+            _traceCategories = traceCategories;
+        }
+
+        internal bool IsAccepted(LogMessage logMessage)
+        {
+            if (_messageTypes.Count == 0 || _messageTypes.Contains(logMessage.Type))
+            {
+                if (logMessage.Type != LogMessageType.TraceMessage || _traceCategories.Count == 0 ||
+                   _traceCategories.Contains(logMessage.TraceCategory))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        internal void Queue(Func<IRemoteLoggerPrx, ValueTask> func, ILogger logger, string op)
+        {
+            lock (this)
```

Comment:
A lock can be used by a single method to prevent concurrent calls, that's what it's doing here, it's easy to find out given that this class has very few users! I'll fix it to use a private mutex. I'll fix LoggerAdmin as well to use a private mutex.